[JSC] Improve `TypeError` message for string methods

sosukesuzuki opened the pull request:
#### 11eba2301506c1c13ad4e95056a3e30801dd90d0
<pre>
[JSC] Improve `TypeError` message for string methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=290454">https://bugs.webkit.org/show_bug.cgi?id=290454</a>

Reviewed by NOBODY (OOPS!).

Currently, functions defined in StringPrototype.js throw
a TypeError with a detailed error message after failing
the coercible check. However, functions defined in
StringPrototype.cpp throw a TypeError without an error message.

This patch changes the functions defined in StringPrototype.cpp
so that they throw a TypeError with a detailed error message for
consistency.

* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::stringIndexOfImpl):
(JSC::toLocaleCase):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11eba2301506c1c13ad4e95056a3e30801dd90d0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96865 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16479 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6666 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101938 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47386 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16775 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24925 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73809 "Found 3 new test failures: js/dom/string-prototype-properties.html js/dom/string-prototype-scopes-in-workers.html js/dom/string-prototype-scopes.html (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31016 "Found 3 new test failures: js/dom/string-prototype-properties.html js/dom/string-prototype-scopes-in-workers.html js/dom/string-prototype-scopes.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99868 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12636 "Found 3 new test failures: js/dom/string-prototype-properties.html js/dom/string-prototype-scopes-in-workers.html js/dom/string-prototype-scopes.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87625 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54143 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12396 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5446 "Found 3 new test failures: js/dom/string-prototype-properties.html js/dom/string-prototype-scopes-in-workers.html js/dom/string-prototype-scopes.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46713 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/89538 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82489 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5527 "Found 3 new test failures: js/dom/string-prototype-properties.html js/dom/string-prototype-scopes-in-workers.html js/dom/string-prototype-scopes.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103962 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/95486 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23933 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17465 "Found 3 new test failures: js/dom/string-prototype-properties.html js/dom/string-prototype-scopes-in-workers.html js/dom/string-prototype-scopes.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82858 "Found 3 new test failures: js/dom/string-prototype-properties.html js/dom/string-prototype-scopes-in-workers.html js/dom/string-prototype-scopes.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24184 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83699 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82249 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26915 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4459 "Found 3 new test failures: js/dom/string-prototype-properties.html js/dom/string-prototype-scopes-in-workers.html js/dom/string-prototype-scopes.html (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17432 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23895 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29050 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119113 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23554 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33484 "Found 16 new JSC stress test failures: stress/string-to-string-error.js.bytecode-cache, stress/string-to-string-error.js.default, stress/string-to-string-error.js.dfg-eager, stress/string-to-string-error.js.dfg-eager-no-cjit-validate, stress/string-to-string-error.js.eager-jettison-no-cjit, stress/string-to-string-error.js.mini-mode, stress/string-to-string-error.js.no-cjit-collect-continuously, stress/string-to-string-error.js.no-llint, stress/string-value-of-error.js.bytecode-cache, stress/string-value-of-error.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27034 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25295 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ef2cf4c91436f3f3d815a28fbbade61d6dd3945a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43040

sosukesuzuki committed:
[JSC] Improve `TypeError` message for string methods
https://bugs.webkit.org/show_bug.cgi?id=290454

Reviewed by NOBODY (OOPS!).

Currently, functions defined in StringPrototype.js throw
a TypeError with a detailed error message after failing
the coercible check. However, functions defined in
StringPrototype.cpp throw a TypeError without an error message.

This patch changes the functions defined in StringPrototype.cpp
so that they throw a TypeError with a detailed error message for
consistency.

* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::stringIndexOfImpl):
(JSC::toLocaleCase):
